WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift

```diff
@@ -1204,23 +1208,48 @@ private extension EditableOrderViewModel {
-              let subTotalDecimal = currencyFormatter.convertToDecimal(orderItem.subtotal),
               let rowViewModel = createProductRowViewModel(for: orderItem, canChangeQuantity: false) else {
             return nil
         }
 
         return ProductInOrderViewModel(productRowViewModel: rowViewModel,
-                                       baseAmountForDiscountPercentage: subTotalDecimal as Decimal,
+                                       productDiscountConfiguration: addProductDiscountConfiguration(on: orderItem),
                                        onRemoveProduct: { [weak self] in
                                             self?.removeItemFromOrder(orderItem)
-                                       },
-                                       onSaveFormattedDiscount: { [weak self] formattedDiscount in
-                                            guard let formattedDiscount = formattedDiscount,
-                                                  let discount = self?.currencyFormatter.convertToDecimal(formattedDiscount) else {
-                                                return
-                                            }
-                                            self?.addDiscountToOrderItem(item: orderItem, discount: discount as Decimal)
-        })
+                                       })
+    }
+
+    /// Creates the configuration related to adding a discount to a product. If the feature shouldn't be shown it returns `nil`
+    ///
+    func addProductDiscountConfiguration(on orderItem: OrderItem) -> ProductInOrderViewModel.DiscountConfiguration? {
+        guard featureFlagService.isFeatureFlagEnabled(.ordersWithCouponsM4),
+              orderSynchronizer.order.coupons.isEmpty,
+              let subTotalDecimal = currencyFormatter.convertToDecimal(orderItem.subtotal) else {
+            return nil
+        }
+
+        return .init(addedDiscount: currentDiscount(on: orderItem),
+                     baseAmountForDiscountPercentage: subTotalDecimal as Decimal,
+                     onSaveFormattedDiscount: { [weak self] formattedDiscount in
+                        guard let formattedDiscount = formattedDiscount,
+                              let discount = self?.currencyFormatter.convertToDecimal(formattedDiscount) else {
+                            self?.addDiscountToOrderItem(item: orderItem, discount: 0)
+                            return
+                        }
+
+                            self?.addDiscountToOrderItem(item: orderItem, discount: discount as Decimal)
+                    })
+    }
+
+    /// Calculates the discount on an order item, that is, subtotal minus total
+    /// 
+    func currentDiscount(on item: OrderItem) -> Decimal {
+        guard let subtotal = currencyFormatter.convertToDecimal(item.subtotal),
+              let total = currencyFormatter.convertToDecimal(item.total) else {
+            return 0
+        }
+
+        return subtotal.subtracting(total) as Decimal
```

Comment:
Would it make sense to extract these two methods to the `ProductInOrderViewModel`? Since this already encapsulates logic for applying discounts to the order it could clean up a bit the `EditableOrderViewModel` and hide implementation details, we can just expose a public method to get the necessary data from it.

Reply:
I would still prefer to leave them in the EditableOrderViewModel for a few reasons:
- We use `currentDiscount` in more cases in `EditableOrderViewModel`
- We would still have to pass the relevant information (e.g. feature flag service, coupons count ...) which would still make it big.
- To keep it consistent, we generate other view models in `EditableOrderViewModel`
- Related to the above, we would still have to pass the logic to add the discount to the synchronizer here, so the closure would be the same. In short, we wouldn't save that much code and would spread the logic in several places, making it more difficult to follow.